### PR TITLE
Enable admin bar menu on site homepage (with valid data)

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Summary

This PR aims to remove the limitation of showing the Admin Bar Site Kit menu only on single post/page URLs. Originally intended to expand adding the menu item only to the home page, I realized as I developed the code that it should now work for **all front end urls**.

Addresses issue https://github.com/google/site-kit-wp/issues/312

## Relevant technical choices

* `Context`: add a `get_reference_canonical` function, returning the current request URL.
* Adjust the `googlesitekit_show_admin_bar_menu` filter to pass the URL (and the current request object when available).
* Improve the logic in `AdminBar::is_active` to only check post status and capability for posts. 
* Only admins can see non post admin bar data. Non post ULRs require `MANAGE_OPTIONS` (admin) capabilities.
* When we can't find a post permalink, fall back to site reference URL plus the current request path (`$wp->request`).
* Admin Bar menu: For the menu's "Stats for:" label - when we can't find a title, fall back to the URL.
* Use `get_reference_canonical` for `Assets` `permaLink` setup.
* Rename `has_data_for_post` to `has_data_for_url`.
* Use a hash of the URL (prefixed for uniqueness) for the transient cache key.
* Data check: Use the last 90 days of data vs. 7 since some URLs may not have data every week.
* Data check: Only check search impressions, not clicks since each click implies an impression.
* Data check: When the URL has data, cache that fact in transient for one day. If not, cache for one hour.

**Additional notes**
* I did quite a bit of testing, however we should still test individual category, tag, custom term pages and archives, author profiles and archives, blog date archives paying special attention to the titles and if possible data (when we have matching public urls with data).
* This URL based approach should work well across front end URLs, however if we want to limit the scope of this change to only show the menu on the home page (and single posts), we can add an `is_home` check in `SearchConsole:: has_data_for_url `.

**Concern** 
* The data check could slow down page loads for users who have the admin bar showing (especially if the api is slow to respond). Should we consider extending cache life to 1 day for "not found" and 1 month for "found"? This status should change much for URLs. 
* Another idea would be using a REST callback to check the data status after the page loads (when its not immediately available from the cache), thus avoiding blocking the load process. the endpoint could also prime the cache.

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit-adminbar.js [hunk 1]
```diff
@@ -80,7 +85,11 @@ export class GoogleSitekitAdminbar extends Component {
 							mdc-layout-grid__cell--align-middle
 						">
 							<span className="googlesitekit-adminbar__subtitle">{ __( 'Stats for:', 'google-site-kit' ) }</span>
-							<h2 className="googlesitekit-adminbar__title">{ googlesitekit.pageTitle && decodeHtmlEntity( googlesitekit.pageTitle ) }</h2>
+							<h2 className="googlesitekit-adminbar__title">
+								{ '' !== pageTitle ?
```
Comment (felixarntz) [inline assets/js/googlesitekit-adminbar.js]: Maybe just to make sure, this could also be `pageTitle && pageTitle.length`?

File: includes/Context.php [hunk 2]
```diff
@@ -197,6 +197,20 @@ public function get_reference_permalink( $post = 0 ) {
 		return $permalink;
 	}
 
+	/**
+	 * Gets the canonical url for the current request.
+	 */
```
Comment (felixarntz) [inline includes/Context.php]: Missing return type doc.

File: includes/Context.php [hunk 3]
```diff
@@ -197,6 +197,20 @@ public function get_reference_permalink( $post = 0 ) {
...
+	public function get_reference_canonical() {
+		global $wp;
+		$reference_permalink = $this->get_reference_permalink();
+		if ( ! $reference_permalink ) {
+
+			// Fall back to site reference URL + current path.
+			return trailingslashit( $this->get_reference_site_url() . $wp->request );
```
Comment (felixarntz) [inline includes/Context.php]: Let's use a more specific URL lookup based on conditions, the current `$wp->request` is not necessarily the canonical. As outlined in the IB, this method should return `false` for anything not (yet) covered.

For now, let's add here:
* If `is_front_page()`, return `$this->get_reference_site_url()`.
* Elseif `is_home()`, return `$this->get_reference_permalink( get_option( 'page_for_posts' ) )`.

We can later expand to more things like, taxonomy archives or post type archives, etc. All of these checks should only be made if `! is_admin()`, because these conditionals are only relevant for the frontend.

Also, because of the more complex need for these conditions, let's reverse the logic a bit to have less nesting, e.g.
```php
if ( $reference_permalink || is_admin() ) {
    return $reference_permalink;
}
```

File: includes/Core/Admin_Bar/Admin_Bar.php [hunk 4]
```diff
@@ -134,41 +134,55 @@ private function add_menu_button( $wp_admin_bar ) {
...
 		}
 
-		// Gets post object. On front area we need to use get_queried_object to get the current post object.
-		if ( $this->is_admin_post_screen() ) {
-			$post = get_post();
-		} else {
-			$post = get_queried_object();
-		}
+		// Determine the queried object.
+		$queried_object = get_queried_object();
```
Comment (felixarntz) [inline includes/Core/Admin_Bar/Admin_Bar.php]: Why was this changed? We should maintain the correct lookup for a single post also in the admin (i.e. block editor).

File: includes/Core/Admin_Bar/Admin_Bar.php [hunk 5]
```diff
@@ -134,41 +134,55 @@ private function add_menu_button( $wp_admin_bar ) {
...
 
-		if ( ! current_user_can( Permissions::VIEW_POST_INSIGHTS, $post->ID ) ) {
-			return false;
+		// Data is based on the current URL.
+		if ( $this->is_admin_post_screen() ) {
+			global $post;
+			$current_url = $this->context->get_reference_permalink( $post->ID );
+		} else {
+			$current_url = $this->context->get_reference_canonical();
 		}
```
Comment (felixarntz) [inline includes/Core/Admin_Bar/Admin_Bar.php]: Based on my previous comment, this clause can be combined with the one for retrieving the current post.

File: includes/Core/Admin_Bar/Admin_Bar.php [hunk 6]
```diff
@@ -134,41 +134,55 @@ private function add_menu_button( $wp_admin_bar ) {
...
+			}
+
+			// Only published posts show the menu.
+			if ( 'publish' !== $queried_object->post_status ) {
+				return false;
+			}
+		} else {
+
+			// Only admins can see non post admin bar data.
+			if ( ! current_user_can( Permissions::MANAGE_OPTIONS ) ) {
```
Comment (felixarntz) [inline includes/Core/Admin_Bar/Admin_Bar.php]: This should better use the (technically similar, but logically more appropriate) `Permissions::VIEW_DASHBOARD` cap.

File: includes/Core/Admin_Bar/Admin_Bar.php [hunk 7]
```diff
@@ -134,41 +134,55 @@ private function add_menu_button( $wp_admin_bar ) {
...
 		 * @since 1.0.0
 		 *
-		 * @param bool $display Whether to display the admin bar menu.
-		 * @param int  $post_id Currently visited post ID.
+		 * @param bool   $display     Whether to display the admin bar menu.
+		 * @param string $current_url The URL of the current request.
+		 * @param mixed  $post_id     Currently queried object if available.
 		 */
-		return apply_filters( 'googlesitekit_show_admin_bar_menu', true, (int) $post->ID );
+		return apply_filters( 'googlesitekit_show_admin_bar_menu', true, $current_url, $queried_object );
```
Comment (felixarntz) [inline includes/Core/Admin_Bar/Admin_Bar.php]: Contrary to what is defined in the ACs, I now actually think we shouldn't pass the post ID here anymore. Whether admin bar stats exist or not entirely depends on the URL. Given that the post can also be `null`, having it as a parameter would be weird anyway.

File: includes/Core/Admin_Bar/Admin_Bar.php [hunk 8]
```diff
@@ -134,41 +134,55 @@ private function add_menu_button( $wp_admin_bar ) {
...
 		}
 
-		if ( ! current_user_can( Permissions::VIEW_POST_INSIGHTS, $post->ID ) ) {
-			return false;
+		// Data is based on the current URL.
+		if ( $this->is_admin_post_screen() ) {
+			global $post;
+			$current_url = $this->context->get_reference_permalink( $post->ID );
+		} else {
+			$current_url = $this->context->get_reference_canonical();
```
Comment (felixarntz) [inline includes/Core/Admin_Bar/Admin_Bar.php]: If this results in `false`, we can return that immediately.

File: includes/Modules/Search_Console.php [hunk 9]
```diff
@@ -362,50 +362,35 @@ protected function create_search_analytics_data_request( array $args = array() )
...
 	 */
-	protected function has_data_for_post( $post_id ) {
-		if ( ! $post_id ) {
+	protected function has_data_for_url( $current_url ) {
+		if ( ! $current_url ) {
 			return false;
 		}
 
-		$transient_key = 'googlesitekit_sc_has_data_for_post_' . $post_id;
+		$transient_key = md5( 'gogolesitekit_sc_data_' . $current_url );
```
Comment (felixarntz) [inline includes/Modules/Search_Console.php]: Two minor issues here:
* A typo in the string.
* Let's only `md5` the `$current_url` and keep the `googlesitekit_sc_data_` prefix in clear text.

File: includes/Modules/Search_Console.php [hunk 10]
```diff
@@ -418,18 +403,20 @@ function( $dataset ) {
 			);
 
 			$has_data = false;
-			foreach ( $responses as $key => $response ) {
-				if ( is_wp_error( $response ) || ! is_array( $response ) || empty( $response ) || ! isset( $response[0] ) ) {
+
+			// Go thru results, any impressions means the URL has data.
+			foreach ( $responses['sc-site-analytics'] as $key => $response ) {
+				if ( is_wp_error( $response ) || empty( $response ) || ! isset( $response ) ) {
```
Comment (felixarntz) [inline includes/Modules/Search_Console.php]: Good catch! This is necessary because of the changed response format right?

File: includes/Modules/Search_Console.php [hunk 11]
```diff
@@ -418,18 +403,20 @@ function( $dataset ) {
...
 
-				if ( $response[0]->clicks > 0 || $response[0]->impressions > 0 ) {
+				if ( $response->impressions > 0 ) {
 					$has_data = true;
 					break;
 				}
 			}
 
-			set_transient( $transient_key, (int) $has_data, 2 * HOUR_IN_SECONDS );
+			set_transient( $transient_key, (int) $has_data, $has_data ? DAY_IN_SECONDS : HOUR_IN_SECONDS );
```
Comment (felixarntz) [inline includes/Modules/Search_Console.php]: Why a different expiration here?

File: assets/js/googlesitekit-adminbar.js (hunk 1 above)
Comment (aaemnnosttv) [reply assets/js/googlesitekit-adminbar.js]: Why not just check for a truthy `pageTitle` since an empty string will be falsy?

File: includes/Context.php (hunk 3 above)
Comment (adamsilverstein) [reply includes/Context.php]: 👍 	Will do, I wasn't sure about using `$wp->request` here.

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 4 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: this was changed in anticipation of other objects we might handle such as a term, comment or user/author. I'll change back to handling only posts explicitly, then we can add other object types later.

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 7 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: I came to the same conclusion while working on this. We really only need the post ID to determine the user's specific level of access to that post (eg a post author might get to see the menu). This is not module/data related, so doesn't belong in the filter.

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 8 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: 👍

File: includes/Modules/Search_Console.php (hunk 9 above)
Comment (adamsilverstein) [reply includes/Modules/Search_Console.php]: 👍  will adjust

File: includes/Modules/Search_Console.php (hunk 11 above)
Comment (adamsilverstein) [reply includes/Modules/Search_Console.php]: when we find data, we really don't need to check for data again for some time (a day). if we don't find data, we can check again more frequently, maybe this is a new site? (1 hour)

File: includes/Modules/Search_Console.php (hunk 10 above)
Comment (adamsilverstein) [reply includes/Modules/Search_Console.php]: Apparently so! When I was testing the `impressions` I was looking for were directly on `$response` not `$response[0]` as in the previous code. My assumption was this had changed in the response and hadn't been adjusted here yet.

File: assets/js/googlesitekit-adminbar.js (hunk 1 above)
Comment (adamsilverstein) [reply assets/js/googlesitekit-adminbar.js]: Adjusted in https://github.com/google/site-kit-wp/pull/543/commits/2b7baf77d072c7ac6b8d5cb5202f88d51868d246

File: includes/Context.php (hunk 2 above)
Comment (adamsilverstein) [reply includes/Context.php]: Added in https://github.com/google/site-kit-wp/pull/543/commits/ba8541758ae39e27bd0abb23085d91a31ed9db74

File: includes/Context.php (hunk 3 above)
Comment (adamsilverstein) [reply includes/Context.php]: Handled in https://github.com/google/site-kit-wp/pull/543/commits/b53eb79f141f17bd1612bf01e0975c65170db1e3

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 4 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: Reverted to previous logic in https://github.com/google/site-kit-wp/pull/543/commits/710095214646bbcf40d63e09f2953f7c7e9fb776

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 5 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: Moved in https://github.com/google/site-kit-wp/pull/543/commits/5166900f0b031a0adad41bbd0cc22d551d5f39e6

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 6 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: adjusted in https://github.com/google/site-kit-wp/pull/543/commits/3794a52aea3030deb4b60a469149c47469b384d9

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 7 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: Removed in https://github.com/google/site-kit-wp/pull/543/commits/2e303bf7bc534523ce1636febdfbc346fd9058a9. Should I update the AC as well?

File: includes/Core/Admin_Bar/Admin_Bar.php (hunk 8 above)
Comment (adamsilverstein) [reply includes/Core/Admin_Bar/Admin_Bar.php]: handled in https://github.com/google/site-kit-wp/pull/543/commits/08c47735b343c4902656e9934eb97d91930c6d43

File: includes/Modules/Search_Console.php (hunk 9 above)
Comment (adamsilverstein) [reply includes/Modules/Search_Console.php]: Adjusted and corrected typo in https://github.com/google/site-kit-wp/pull/543/commits/fce143f6c305633b2250727e9a6cb1f9db3f691e

**Comment by adamsilverstein** [comment]
@felixarntz Thank you for the detailed feedback. I have addressed each point in a discreet commit and re-tested with the final result, verifying data checks work and I see the site kit admin bar menu on the home page, dataful url and wp-admin editor editing a page with data.

File: includes/Context.php [hunk 12]
```diff
@@ -164,11 +164,11 @@ public function get_reference_site_url() {
...
 	 * @since 1.0.0
 	 *
-	 * @param int|\WP_Post $post  Optional. Post ID or post object. Default is the global `$post`.
+	 * @param int|\WP_Post $post  Post ID or post object. Optional. Default is the global `$post`.
 	 *
-	 * @return string|false The  Reference permalink URL or false if post does not exist.
+	 * @return string|false The reference permalink URL or false if post does not exist.
 	 */
-	public function get_reference_permalink( $post = 0 ) {
+	public function get_reference_permalink( $post = false ) {
```
Comment (adamsilverstein) [inline includes/Context.php]: setting as `false` by default, 0 feels like the wrong value here

File: includes/Core/Admin_Bar/Admin_Bar.php [hunk 13]
```diff
@@ -134,41 +134,57 @@ private function add_menu_button( $wp_admin_bar ) {
 	 * @return bool True if Admin bar should display, False when it's not.
 	 */
 	public function is_active() {
-		if ( ! is_user_logged_in() || ! is_admin_bar_showing() ) {
+
+		// Only active if the admin bar is showing.
+		if ( ! is_admin_bar_showing() ) {
```
Comment (adamsilverstein) [inline includes/Core/Admin_Bar/Admin_Bar.php]: note: `is_admin_bar_showing` checks `is_user_logged_in` so i removed that here

File: includes/Context.php [hunk 14]
```diff
@@ -164,11 +164,11 @@ public function get_reference_site_url() {
 	 *
 	 * @since 1.0.0
 	 *
-	 * @param int|\WP_Post $post  Optional. Post ID or post object. Default is the global `$post`.
+	 * @param int|\WP_Post $post  Post ID or post object. Optional. Default is the global `$post`.
```
Comment (felixarntz) [inline includes/Context.php]: Per WP core standards, the `Optional.` piece should be the beginning of optional parameter descriptions, so this should be changed back.

File: includes/Context.php (hunk 12 above)
Comment (felixarntz) [reply includes/Context.php]: Why that? `0` is also "falsey", and `false` does not match the method signature. I think this should remain as it is.

File: includes/Context.php (hunk 14 above)
Comment (aaemnnosttv) [reply includes/Context.php]:
```suggestion
	 * @param int|\WP_Post $post  Optional. Post ID or post object. Default is the global `$post`.
```

File: includes/Context.php (hunk 12 above)
Comment (aaemnnosttv) [reply includes/Context.php]:
```suggestion
	public function get_reference_permalink( $post = 0 ) {
```
